plugin/trino-sqlserver/src/main/java/io/trino/plugin/sqlserver/SqlServerStatsCollector.java

```diff
@@ -0,0 +1,440 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.sqlserver;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import com.google.inject.Inject;
+import io.airlift.log.Logger;
+import io.trino.plugin.jdbc.ConnectionFactory;
+import io.trino.plugin.jdbc.JdbcColumnHandle;
+import io.trino.plugin.jdbc.JdbcTableHandle;
+import io.trino.plugin.jdbc.RemoteTableName;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.statistics.ColumnStatistics;
+import io.trino.spi.statistics.Estimate;
+import io.trino.spi.statistics.TableStatistics;
+import org.jdbi.v3.core.Handle;
+import org.jdbi.v3.core.Jdbi;
+
+import java.sql.CallableStatement;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.MoreCollectors.toOptional;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class SqlServerStatsCollector
+{
+    private static final Logger log = Logger.get(SqlServerStatsCollector.class);
+
+    private final ConnectionFactory connectionFactory;
+
+    @Inject
+    public SqlServerStatsCollector(ConnectionFactory connectionFactory)
+    {
+        this.connectionFactory = requireNonNull(connectionFactory, "connectionFactory is null");
+    }
+
+    TableStatistics readTableStatistics(ConnectorSession session, JdbcTableHandle table, SqlServerClient sqlServerClient)
+            throws SQLException
+    {
+        checkArgument(table.isNamedRelation(), "Relation is not a table: %s", table);
+
+        if (isNewStatisticsSupported(session)) {
```

Comment:
Can we optimistically try reading new stats, detect a table-not-found exception and fall back to legacy stats? We could also memoize this.

Reply:
Not sure why we'd get a table-not-found exception in new stats but not legacy stats, but regardless, this is a draft PR and i've only been trying to reach parity - working on a fallback like that is premature at this stage.